### ISSUE TITLE
Wrong [7F] in spherical molden files

### ISSUE THREAD

**Issue by albandil** [psi4/psi4#2612]
I believe that the issue https://github.com/psi4/psi4/issues/2238 was misleading.

When I produce a Molden file in self-compiled Psi4 1.6 using this code
```
molecule {
    0  1
    He
}

set basis cc-pVTZ

[E, wfn] = energy('scf', return_wfn=True)

wfn.write_molden('He.molden')
```
the flags written to the file are
```
[5D]
[7F]
[9G]
```
However, this seems to be in conflict with the [Molden format specification](https://www.theochem.ru.nl/molden/molden_format.html), where `[7F]` means "Cartesian d-functions and spherical f-functions". Also, the *HISTORY* file from Molden 7.1 distribution contains the following summary, dated to Molden 4.1:
```
                  - default is    6D and 10F
                  - [5D7F]  means 5D and  7F
                  - [5D10F] means 5D and 10F
                  - [7F]    means 6D and  7F

                  for backwards compatibility, [5D] still means 5D and  7F.
```
See also this excerpt from Molden 7.1, subroutine `prtmold` (output of a Molden file from Molden):
```fortran
         if (ido5d.eq.1.and.ido7f.eq.1) then
             write(iun3,'(a)') '[5D7F]'
         elseif (ido5d.eq.1.and.ido7f.eq.0) then
             write(iun3,'(a)') '[5D10F]'
         elseif (ido5d.eq.0.and.ido7f.eq.1) then
             write(iun3,'(a)') '[7F]'
         elseif (ido9g.eq.1) then
             write(iun3,'(a)') '[9G]'
         endif
```
and the subroutine `rdmodd` (input of a Molden file into Molden):
```fortran
            call srchmf(line,'[5D',istat)
            if (istat.eq.1) then
               if (icdex(line,'[5D]').ne.0.or.
     &             icdex(line,'[5D7F]').ne.0) then
                   ido5d = 1
                   ido7f = 1
               elseif (icdex(line,'[5D10F]').ne.0) then
                   ido5d = 1
               endif
            endif
            call rewmf
            call srchmf(line,'[7F]',istat)
            if (istat.eq.1) ido7f = 1
            call rewmf
            call srchmf(line,'[9G]',istat)
            if (istat.eq.1) ido9g = 1

```
To sum up, it appears to me that a correct Molden file
 - no longer uses `[5D]` in the first place,
 - uses `[7F]` to mean "Cartesian d-functions, spherical f-functions" and
 - uses `[5D7F]` and `[5D10F]` for spherical d-functions and Cartesian or spherical f-functions, respectively.

When I use the legacy writer
```
molden(wfn, 'He.molden')
```
the output contains `[5D]` and `[9G]`, which at least seems to fall into the correct "backwards compatibility" mentioned in the *HISTORY* file.

**Comment by JonathonMisiewicz** [comment]
For all spherical, I believe the correct tag would then be `[5D7F9G]`?

**Comment by albandil** [comment]
For all spherical, there will be two separate tags: `[5D7F]` (or just `[5D]` in the "backward compatibility" mode) and `[9G]`.

**Comment by JonathonMisiewicz** [comment]
The Molden documentation is unclear, but based on the code snippets you posted, your understanding of Molden's behavior is in error, and Psi is correct.

1. `rdmodd` is proof-positive that `[5D]` is used. For historical reasons, it means "use spherical D and F functions". This is consistent with what the Molden documentation says: "Use the keyword [5D] on a separate line to specify the use of 'spherical' D and F functions (5 D and 7 F functions)." While I find the meaning of the keyword unintuitive and confusing, I see nothing to support your assessment that the use of `[5D]` is _incorrect_.
2. `rdmodd` shows that `[7F]` uses spherical f-functions but makes no statement whatsoever about the `d` functions. While the documentation says that keywords such as `[7F]` enable the use of mixed spherical and cartesians for the d and f, the documentation does not specify whether `[7F]` does this by implicitly assuming `6D` or by not changing the `D` at all, which may well be in Cartesians.
3. Your third bullet point seem to have been miswritten. You want "spherical or Cartesian" and not "Cartesian or spherical".

By specifying [5D], [7F], and [9G], Psi gives Molden the following instructions:
* Use spherical D and F shells.
* Use spherical F shells. (This is redundant but harmless, and therefore, your issue is in error.)
* Use spherical G shells.

I'll leave the issue open in case you point out something I've overlooked.

**Comment by albandil** [comment]
My first point was that Molden hasn't been using `[5D]` anymore for some time when writing Molden files (since 2004). It might be a good idea to follow it to avoid ambiguities.

My second point is that software that follows Molden format specification (I am developing one such code) will likely see a conflict in `[5D]` and `[7F]`. Molden does not use these two flags together when writing Molden files and it never did. Even though it attempts to silently interpret their combination in some way, `[7F]` was introduced in the same version (4.1) when `[5D]` was abandoned.

You are right about the interchanged words in the third point.

Thank you for consideration.

**Comment by JonathonMisiewicz** [comment]
That Psi's Molden files _are not using best practices_ is an entirely different matter than _are wrong_. I can change this with two lines of code, but be more careful when filing issues in future. The issue as filed was factually incorrect in some places and needlessly confusing in others.

**Comment by loriab** [comment]
The missing words in the main Molden docs make sph/cart flags difficult to reason about with confidence, so thanks for digging up the changelog and filing a report, @albandil. Thanks for the fix, @JonathonMisiewicz.
